pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+            self.set_idata_attrs(prior_pred)
+
+        if extend_idata:
+            if self.idata is not None:
+                self.idata.extend(prior_pred)
+            else:
+                self.idata = prior_pred
+
+    def fit(self, extend_idata, kwargs):
+        """Fit Nested Logit Model."""
+        if extend_idata:
+            with self.model:
+                self.idata.extend(pm.sample(**kwargs))
+        else:
+            with self.model:
+                self.idata = pm.sample(**kwargs)
+
+    def sample_posterior_predictive(self, extend_idata, kwargs):
+        """Sample Posterior Predictive Distribution."""
+        if extend_idata:
+            with self.model:
+                self.idata.extend(
+                    pm.sample_posterior_predictive(
+                        self.idata, var_names=["likelihood", "p"], **kwargs
+                    )
+                )
+        else:
+            with self.model:
+                self.post_pred = pm.sample_posterior_predictive(
+                    self.idata, var_names=["likelihood", "p"], **kwargs
+                )
+
+    def sample(
+        self,
+        sample_prior_predictive_kwargs: dict | None = None,
+        fit_kwargs: dict | None = None,
+        sample_posterior_predictive_kwargs: dict | None = None,
+    ) -> Self:
+        """Sample all the things.
+
+        Run all of the sample methods in the sequence:
+
+        - :meth:`sample_prior_predictive`
+        - :meth:`fit`
+        - :meth:`sample_posterior_predictive`
+
+        Parameters
+        ----------
+        sample_prior_predictive_kwargs : dict, optional
+            The keyword arguments for the sample_prior_predictive method.
+        fit_kwargs : dict, optional
+            The keyword arguments for the fit method.
+        sample_posterior_predictive_kwargs : dict, optional
+            The keyword arguments for the sample_posterior_predictive method.
+
+        Returns
+        -------
+        Self
+            The model instance.
+
+        """
+        sample_prior_predictive_kwargs = sample_prior_predictive_kwargs or {}
+        fit_kwargs = fit_kwargs or {
+            "nuts_sampler": "numpyro",
+            "idata_kwargs": {"log_likelihood": True},
+        }
+        sample_posterior_predictive_kwargs = sample_posterior_predictive_kwargs or {}
+        if not hasattr(self, "model"):
+            X, F, y = self.preprocess_model_data(self.choice_df, self.utility_equations)  # type: ignore
+            model = self.make_model(X, F, y)
+            self.model = model
+
+        self.sample_prior_predictive(
+            extend_idata=True, kwargs=sample_prior_predictive_kwargs
+        )
+        self.fit(extend_idata=True, kwargs=fit_kwargs)
+        self.sample_posterior_predictive(
+            extend_idata=True, kwargs=sample_posterior_predictive_kwargs
+        )
+        return self
+
+    def apply_intervention(self, new_choice_df, new_utility_equations=None):
```

Comment:
type hints missing and parameters in docstings

Reply:
added